File: src/org/genyris/io/writerstream/WriterStream.java
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.io.writerstream;

import java.io.IOException;
import java.io.Writer;

import org.genyris.core.Atom;
import org.genyris.core.Constants;
import org.genyris.core.Exp;
import org.genyris.core.Internable;
import org.genyris.core.StrinG;
import org.genyris.core.Symbol;
import org.genyris.core.Visitor;
import org.genyris.exception.GenyrisException;
import org.genyris.format.BasicFormatter;
import org.genyris.format.DisplayFormatter;
import org.genyris.format.Formatter;
import org.genyris.format.HTMLFormatter;
import org.genyris.format.JSONFormatter;
import org.genyris.format.UrlFormatter;
import org.genyris.interp.AbstractMethod;
import org.genyris.interp.Closure;
import org.genyris.interp.Environment;
import org.genyris.interp.Interpreter;
import org.genyris.interp.UnboundException;

public class WriterStream extends Atom {
    private Writer _value;

    public WriterStream(Writer w) {
        _value = w;
    }

    public void acceptVisitor(Visitor guest) throws GenyrisException {
        guest.visitExpWithEmbeddedClasses(this);
    }

    public String toString() {
        return _value.toString();
    }

    public Symbol getBuiltinClassSymbol(Internable table) {
        return table.WRITER();
    }

    public void write(char ch) throws GenyrisException {
        try {
            _value.write(ch);
        }
        catch (IOException e) {
            throw new GenyrisException(e.getMessage());
        }
    }

    public void close() throws GenyrisException {
        try {
            _value.close();
        }
        catch (IOException e) {
            throw new GenyrisException(e.getMessage());
        }
    }

    public Exp format(StrinG formatString, int offset, Exp[] args, Environment env) throws GenyrisException {
        StringBuffer format = new StringBuffer(formatString.toString());
        int argCounter = offset;
        char escape = '%';
        try {
            for (int i = 0; i < format.length(); i++) {
                if ((format.charAt(i) == escape) && (i == format.length()-1)) {
                	throw new GenyrisException("Bad format: " + format);
                }
                if (format.charAt(i) == escape && format.charAt(i + 1) == 'a') {
                    // display - TODO DRY
                    i++;
                    if (argCounter > args.length) {
                        break;
                    }
                    Formatter formatter = new DisplayFormatter(_value);
                    if(argCounter == args.length) {
                        throw new GenyrisException("Bad format: " + format + " too few real arguments.");                    	
                    }
                    args[argCounter++].acceptVisitor(formatter);
                } else if (format.charAt(i) == escape && format.charAt(i + 1) == 's') {
                    // write - TODO DRY
                    i++;
                    if (argCounter > args.length) {
                        break;
                    }
                    Formatter formatter = new BasicFormatter(_value);
                    if(argCounter == args.length) {
                        throw new GenyrisException("Bad format: " + format + " too few real arguments.");                    	
                    }
                    args[argCounter++].acceptVisitor(formatter);
                } else if (format.charAt(i) == escape && format.charAt(i + 1) == 'x') {
                    // write - TODO DRY
                    i++;
                    if (argCounter > args.length) {
                        break;
                    }
                    Formatter formatter = new HTMLFormatter(_value);
                    if(argCounter == args.length) {
                        throw new GenyrisException("Bad format: " + format + " too few real arguments.");                    	
                    }
                    args[argCounter++].acceptVisitor(formatter);
                } else if (format.charAt(i) == escape && format.charAt(i + 1) == 'u') {
                    // write - TODO DRY
                    i++;
                    if (argCounter > args.length) {
                        break;
                    }
                    Formatter formatter = new UrlFormatter(_value);
                    if(argCounter == args.length) {
                        throw new GenyrisException("Bad format: " + format + " too few real arguments.");                    	
                    }
                    args[argCounter++].acceptVisitor(formatter);
                } else if (format.charAt(i) == escape && format.charAt(i + 1) == 'n') {
                    i++;
                    _value.append('\n');
                } else if (format.charAt(i) == escape && format.charAt(i + 1) == 'j') {
                    i++;
                    if (argCounter > args.length) {
                        break;
                    }
                    Formatter formatter = new JSONFormatter(_value);
                    if(argCounter == args.length) {
                        throw new GenyrisException("Bad format: " + format + " too few real arguments.");
                    }
                    args[argCounter++].acceptVisitor(formatter);
                } else if (format.charAt(i) == escape && format.charAt(i + 1) == escape) {
                    i++;
                    _value.append(escape);
                } else {
                    _value.append(format.charAt(i));
                }
            }
            if(argCounter != args.length) {
                throw new GenyrisException("Bad format: " + format + " too many real arguments.");                    	            	
            }
        }
        catch (ArrayIndexOutOfBoundsException e) {
            throw new GenyrisException("Bad format: " + format + " " + e.getMessage());
        }
        catch (StringIndexOutOfBoundsException e) {
            throw new GenyrisException("Bad format: " + format + " " + e.getMessage());
        }
        catch (IOException e) {
        	if(e.getMessage() == null) {
                throw new GenyrisException("IOException");        		
        	}
            throw new GenyrisException(e.getMessage());
        }
        return env.getNil();
    }
    public static abstract class AbstractWriterMethod extends AbstractMethod {

        public AbstractWriterMethod(Interpreter interp, String name) {
            super(interp, name);
        }

        protected WriterStream getSelfWriter(Environment env) throws GenyrisException {
            getSelf(env);
            if (!(_self instanceof WriterStream)) {
                throw new GenyrisException("Non-Writer passed to a Writer method.");
            } else {
                return (WriterStream)_self;
            }
        }
        public static void bindFunctionsAndMethods(Interpreter interpreter) throws UnboundException, GenyrisException {
            interpreter.bindMethodInstance(Constants.WRITER, new FormatMethod(interpreter));
            interpreter.bindMethodInstance(Constants.WRITER, new CloseMethod(interpreter));
            interpreter.bindMethodInstance(Constants.WRITER, new FlushMethod(interpreter));
        }
    }
    public static class FormatMethod extends AbstractWriterMethod {
        private Exp STDOUT;
        private Exp STDERR;

        public static String getStaticName() {return "format";};
        public FormatMethod(Interpreter interp) {
            super(interp, getStaticName());
            try {
                STDOUT = interp.lookupGlobalFromString(Constants.STDOUT);
                STDERR = interp.lookupGlobalFromString(Constants.STDERR);
            }
            catch (GenyrisException e) {
                STDOUT = null;
                STDERR = null;
            }
        }

        public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
                throws GenyrisException {
            if (arguments.length > 0) {
                if (!(arguments[0] instanceof StrinG)) {
                    throw new GenyrisException("Non string passed to FormatMethod");
                }
                WriterStream self = getSelfWriter(env);
                Exp retval = self.format((StrinG)arguments[0], 1, arguments, env);
                if(self == STDOUT | self == STDERR) {
                    try {
                        self._value.flush();
                    }
                    catch (IOException e) {
                        throw new GenyrisException(e.getMessage());
                    }
                }
                return retval;
            } else {
                throw new GenyrisException("Missing argument to FormatMethod");
            }
        }
    }
    public static class CloseMethod extends AbstractWriterMethod {

        public static String getStaticName() {return "close";};
        public CloseMethod(Interpreter interp) {
            super(interp, getStaticName());
        }

        public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
                throws GenyrisException {
            getSelfWriter(env).close();
            return NIL;
        }
    }
    public static class FlushMethod extends AbstractWriterMethod {

        public static String getStaticName() {return "flush";};
        public FlushMethod(Interpreter interp) {
            super(interp, getStaticName());
        }

        public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
                throws GenyrisException {
                getSelfWriter(env).flush();
                return NIL;
        }
    }
    public Exp flush() throws GenyrisException {
        try {
            _value.flush();
        }
        catch (IOException e) {
            throw new GenyrisException(e.getMessage());
        }
        return null;
    }
	public Exp eval(Environment env) throws GenyrisException {
		return this;
	}

	public Writer getWriter() {
		return _value;
	}

}

File: src/org/genyris/io/writerstream/WriterStream.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.io.writerstream

import org.genyris.core.*
import org.genyris.exception.GenyrisException
import org.genyris.format.*
import org.genyris.interp.AbstractMethod
import org.genyris.interp.Closure
import org.genyris.interp.Environment
import org.genyris.interp.Interpreter
import java.io.IOException
import java.io.Writer

class WriterStream(val writer: Writer) : Atom(), Comparable<Any?> {
    @kotlin.Throws(GenyrisException::class)
    override fun acceptVisitor(guest: Visitor) {
        guest.visitExpWithEmbeddedClasses(this)
    }

    override fun toString(): String {
        return writer.toString()
    }

    override fun getBuiltinClassSymbol(table: Internable): Symbol? {
        return table.WRITER()
    }

    @kotlin.Throws(GenyrisException::class)
    fun write(ch: Char) {
        try {
            writer.write(ch.code)
        } catch (e: IOException) {
            throw GenyrisException(e.getMessage())
        }
    }

    @kotlin.Throws(GenyrisException::class)
    fun close() {
        try {
            writer.close()
        } catch (e: IOException) {
            throw GenyrisException(e.getMessage())
        }
    }

    @kotlin.Throws(GenyrisException::class)
    fun format(formatString: StrinG, offset: Int, args: Array<Exp?>, env: Environment): Exp? {
        val format = StringBuffer(formatString.toString())
        var argCounter = offset
        val escape = '%'
        try {
            var i = 0
            while (i < format.length()) {
                val ch: Char = format.charAt(i)
                if (ch != escape) {
                    writer.append(ch)
                    i++
                    continue
                }
                if ((ch == escape) && (i == format.length() - 1)) {
                    throw GenyrisException("Bad format: " + format)
                }
                val order: Char = format.charAt(i + 1)
                if (ch == escape && order == 'a') {
                    // display - TODO DRY
                    i++
                    if (argCounter > args.size) {
                        break
                    }
                    val formatter: Formatter = DisplayFormatter(this.writer)
                    if (argCounter == args.size) {
                        throw GenyrisException("Bad format: " + format + " too few real arguments.")
                    }
                    args[argCounter++]!!.acceptVisitor(formatter)
                } else if (ch == escape && (order == 'i' || order == 'I')) {
                    // write - TODO DRY
                    i++
                    if (argCounter > args.size) {
                        break
                    }
                    val formatter: Formatter =
                        IndentedFormatter(this.writer, 3, (if (Character.isUpperCase(order)) true else false))
                    if (argCounter == args.size) {
                        throw GenyrisException("Bad format: " + format + " too few real arguments.")
                    }
                    args[argCounter++]!!.acceptVisitor(formatter)
                } else if (ch == escape && (order == 's' || order == 'S')) {
                    // write - TODO DRY
                    if (argCounter > args.size) {
                        break
                    }
                    val formatter: Formatter =
                        BasicFormatter(this.writer, (if (Character.isUpperCase(order)) true else false))
                    if (argCounter == args.size) {
                        throw GenyrisException("Bad format: " + format + " too few real arguments.")
                    }
                    i++
                    args[argCounter++]!!.acceptVisitor(formatter)
                } else if (ch == escape && order == 'x') {
                    // write - TODO DRY
                    i++
                    if (argCounter > args.size) {
                        break
                    }
                    val formatter: Formatter = HTMLFormatter(this.writer)
                    if (argCounter == args.size) {
                        throw GenyrisException("Bad format: " + format + " too few real arguments.")
                    }
                    args[argCounter++]!!.acceptVisitor(formatter)
                } else if (ch == escape && order == 'u') {
                    // write - TODO DRY
                    i++
                    if (argCounter > args.size) {
                        break
                    }
                    val formatter: Formatter = UrlFormatter(this.writer)
                    if (argCounter == args.size) {
                        throw GenyrisException("Bad format: " + format + " too few real arguments.")
                    }
                    args[argCounter++]!!.acceptVisitor(formatter)
                } else if (ch == escape && order == 'n') {
                    i++
                    writer.append('\n')
                } else if (ch == escape && order == 'j') {
                    i++
                    if (argCounter > args.size) {
                        break
                    }
                    val formatter: Formatter = JSONFormatter(this.writer)
                    if (argCounter == args.size) {
                        throw GenyrisException("Bad format: " + format + " too few real arguments.")
                    }
                    args[argCounter++]!!.acceptVisitor(formatter)
                } else if (ch == escape && order == escape) {
                    i++
                    writer.append(escape)
                } else {
                    writer.append(ch)
                }
                i++
            }
            if (argCounter != args.size) {
                throw GenyrisException("Bad format: " + format + " too many real arguments.")
            }
        } catch (e: ArrayIndexOutOfBoundsException) {
            throw GenyrisException("Bad format: " + format + " " + e.getMessage())
        } catch (e: StringIndexOutOfBoundsException) {
            throw GenyrisException("Bad format: " + format + " " + e.getMessage())
        } catch (e: IOException) {
            if (e.getMessage() == null) {
                throw GenyrisException("IOException")
            }
            throw GenyrisException(e.getMessage())
        }
        return env.getNil()
    }

    abstract class AbstractWriterMethod(interp: Interpreter?, name: String?) : AbstractMethod(interp, name) {
        @kotlin.Throws(GenyrisException::class)
        protected fun getSelfWriter(env: Environment): WriterStream {
            getSelf(env)
            if (_self !is WriterStream) {
                throw GenyrisException("Non-Writer passed to a Writer method.")
            } else {
                return _self as WriterStream
            }
        }

        companion object {
            @kotlin.Throws(GenyrisException::class)
            fun bindFunctionsAndMethods(interpreter: Interpreter) {
                interpreter.bindMethodInstance(Constants.WRITER, FormatMethod(interpreter))
                interpreter.bindMethodInstance(Constants.WRITER, CloseMethod(interpreter))
                interpreter.bindMethodInstance(Constants.WRITER, FlushMethod(interpreter))
            }
        }
    }

    class FormatMethod(interp: Interpreter) : AbstractWriterMethod(
        interp,
        staticName
    ) {
        private var STDOUT: Exp? = null
        private var STDERR: Exp? = null

        init {
            try {
                STDOUT = interp.lookupGlobalFromString(Constants.STDOUT)
                STDERR = interp.lookupGlobalFromString(Constants.STDERR)
            } catch (e: GenyrisException) {
                STDOUT = null
                STDERR = null
            }
        }

        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp?>, env: Environment): Exp? {
            if (arguments.size > 0) {
                if (arguments[0] !is StrinG) {
                    throw GenyrisException("Non string passed to FormatMethod")
                }
                val self = getSelfWriter(env)
                val retval = self.format((arguments[0] as StrinG?)!!, 1, arguments, env)
                if ((self === STDOUT) or (self === STDERR)) {
                    try {
                        self.writer.flush()
                    } catch (e: IOException) {
                        throw GenyrisException(e.getMessage())
                    }
                }
                return retval
            } else {
                throw GenyrisException("Missing argument to FormatMethod")
            }
        }

        companion object {
            val staticName: String
                get() = "format"
        }
    }

    class CloseMethod(interp: Interpreter?) : AbstractWriterMethod(
        interp,
        staticName
    ) {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp?>?, env: Environment): Exp? {
            getSelfWriter(env).close()
            return NIL
        }

        companion object {
            val staticName: String
                get() = "close"
        }
    }

    class FlushMethod(interp: Interpreter?) : AbstractWriterMethod(
        interp,
        staticName
    ) {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp?>?, env: Environment): Exp? {
            getSelfWriter(env).flush()
            return NIL
        }

        companion object {
            val staticName: String
                get() = "flush"
        }
    }

    @kotlin.Throws(GenyrisException::class)
    fun flush(): Exp? {
        try {
            writer.flush()
        } catch (e: IOException) {
            throw GenyrisException(e.getMessage())
        }
        return null
    }

    @kotlin.Throws(GenyrisException::class)
    override fun eval(env: Environment?): Exp {
        return this
    }

    override fun compareTo(o: Any?): Int {
        return if (this === o) 0 else 1
    }
}
